booking/src/main/java/org/dcsa/conformance/standards/booking/checks/BookingChecks.java

```diff
@@ -0,0 +1,645 @@
+package org.dcsa.conformance.standards.booking.checks;
+
+import com.fasterxml.jackson.core.JsonPointer;
+import com.fasterxml.jackson.databind.JsonNode;
+import lombok.experimental.UtilityClass;
+import org.dcsa.conformance.core.check.*;
+import org.dcsa.conformance.core.traffic.HttpMessageType;
+import org.dcsa.conformance.standards.booking.party.BookingRole;
+import org.dcsa.conformance.standards.booking.party.BookingState;
+import org.dcsa.conformance.standards.booking.party.CarrierScenarioParameters;
+import org.dcsa.conformance.standards.booking.party.DynamicScenarioParameters;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static org.dcsa.conformance.standards.booking.checks.BookingDataSets.CUTOFF_DATE_TIME_CODES;
+
+@UtilityClass
+public class BookingChecks {
+
+  private static final Set<BookingState> CONFIRMED_BOOKING_STATES = Set.of(
+    BookingState.CONFIRMED,
+    BookingState.PENDING_AMENDMENT,
+    BookingState.COMPLETED,
+    BookingState.DECLINED
+  );
+
+  private static final Set<BookingState> PENDING_CHANGES_STATES = Set.of(
+    BookingState.PENDING_UPDATE,
+    BookingState.PENDING_AMENDMENT
+  );
+
+  private static final Set<BookingState> REASON_STATES = Set.of(
+    BookingState.DECLINED,
+    BookingState.REJECTED,
+    BookingState.CANCELLED,
+    BookingState.AMENDMENT_CANCELLED,
+    BookingState.AMENDMENT_DECLINED
+  );
+
+  private static final Set<BookingState> BOOKING_STATES_WHERE_CBR_IS_OPTIONAL = Set.of(
+    BookingState.RECEIVED,
+    BookingState.REJECTED,
+    BookingState.PENDING_UPDATE,
+    BookingState.UPDATE_RECEIVED,
+    /* CANCELLED depends on whether cancel happens before CONFIRMED, but the logic does not track prior
+     * states. Therefore, we just assume it is optional in CANCELLED here.
+     */
+    BookingState.CANCELLED
+  );
+  private static final JsonPointer CARRIER_BOOKING_REQUEST_REFERENCE = JsonPointer.compile("/carrierBookingRequestReference");
+  private static final JsonPointer CARRIER_BOOKING_REFERENCE = JsonPointer.compile("/carrierBookingReference");
+  private static final JsonPointer BOOKING_STATUS = JsonPointer.compile("/bookingStatus");
+  private static final JsonPointer[] TD_UN_LOCATION_CODES = {
+    JsonPointer.compile("/invoicePayableAt/UNLocationCode"),
+    JsonPointer.compile("/shipmentLocations/location/UNLocationCode"),
+    JsonPointer.compile("/shipmentLocations/location/facilityLocation/UNLocationCode")
```

Comment:
This does not work; `shipmentLocations` is an array and JSON pointer can only match a single node. In the concrete case, the pointer matches nothing (since the path would have be /shipmentLocations/0/... + /shipmentLocations/1/... + ..., which will not scale).

You would have to use a different approach when arrays are involved.